pkg/actuators/machine/instances.go

```diff
@@ -373,57 +373,66 @@ func getImageID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabac
+	klog.Infof("query security groups in region %s", machineProviderConfig.RegionID)
 	var securityGroupIDs []string
 
 	// If SecurityGroupID is assigned, use it directly
-	if machineProviderConfig.SecurityGroupID != "" {
-		securityGroupIDs = append(securityGroupIDs, machineProviderConfig.SecurityGroupID)
-	} else {
-		// Otherwise, the query securityGroupIDs by the tags
-		for _, sg := range machineProviderConfig.SecurityGroups {
-			if sg.ID != "" {
-				securityGroupIDs = append(securityGroupIDs, sg.ID)
-			} else {
-				if sg.Tags != nil {
-					request := ecs.CreateDescribeSecurityGroupsRequest()
-					if machineProviderConfig.VpcID != "" {
-						request.VpcId = machineProviderConfig.VpcID
-					}
-					if machineProviderConfig.ResourceGroupID != "" {
-						request.ResourceGroupId = machineProviderConfig.ResourceGroupID
-					}
-					request.RegionId = machineProviderConfig.RegionID
-					request.Tag = buildDescribeSecurityGroupsTag(sg.Tags)
-					request.Scheme = "https"
-
-					response, err := client.DescribeSecurityGroups(request)
-					if err != nil {
-						metrics.RegisterFailedInstanceCreate(&metrics.MachineLabels{
-							Name:      machine.Name,
-							Namespace: machine.Namespace,
-							Reason:    err.Error(),
-						})
-						klog.Errorf("error describing securitygroup: %v", err)
-						return nil, fmt.Errorf("error describing securitygroup: %v", err)
-					}
-
-					if len(response.SecurityGroups.SecurityGroup) < 1 {
-						klog.Errorf("no securitygroup for given tags not found")
-						return nil, fmt.Errorf("no securitygroup for given tags not found")
-					}
-
-					for _, sg := range response.SecurityGroups.SecurityGroup {
-						securityGroupIDs = append(securityGroupIDs, sg.SecurityGroupId)
-					}
+	if machineProviderConfig.SecurityGroups == nil {
+		return nil, errors.New("no security configuration provided")
+	}
+
+	for _, sg := range machineProviderConfig.SecurityGroups {
+		if sg.ID != "" {
+			securityGroupIDs = append(securityGroupIDs, sg.ID)
+		} else {
+			if sg.Tags != nil {
+				ids, err := getSecurityGroupIDByTags(machine, machineProviderConfig, sg.Tags, client)
+				if err != nil {
+					return nil, err
 				}
+				securityGroupIDs = append(securityGroupIDs, ids...)
 			}
 		}
 	}
-
+	if len(securityGroupIDs) == 0 {
+		return nil, errors.New("no securitygroup IDs found from configuration")
+	}
 	return &securityGroupIDs, nil
 }
 
+func getSecurityGroupIDByTags(machine runtimeclient.ObjectKey, machineProviderConfig *alibabacloudproviderv1.AlibabaCloudMachineProviderConfig, tags []alibabacloudproviderv1.Tag, client alibabacloudClient.Client) ([]string, error) {
+	request := ecs.CreateDescribeSecurityGroupsRequest()
+	if machineProviderConfig.VpcID != "" {
+		request.VpcId = machineProviderConfig.VpcID
+	}
+	if machineProviderConfig.ResourceGroupID != "" {
+		request.ResourceGroupId = machineProviderConfig.ResourceGroupID
+	}
```

Comment:
Do these need to be behind if statements?

Reply:
I honestly do not know. It appears in the `DescribeSecurityGroups` API call that when making the request the VpcId and the ResourceGroupID looks like this:
```
 VpcId: (string) ""
 ResourceGroupId: (string) (len=18) "rg-acfnxrgr6qtm3mq",
```
so I think you are probably on to something. Setting them to whatever the default is on the machineProviderConfig would probably work since it is defaulting them to empty anyway.